drafts/two-way-binding-signals-and-query-params.md

```diff
@@ -0,0 +1,754 @@
+    this.router.events.subscribe((e) => console.log(e));
+  }
+
+  // Same as before
+  get(queryParamName: string) {
+    const activatedRouteValue = computed(() => this.allQueryParams()?.[queryParamName]);
+    const queryParamValue = signal(activatedRouteValue());
+
+    effect(() => queryParamValue.set(activatedRouteValue()), { allowSignalWrites: true });
+
+    effect(() =>
+      this.router.navigate([], {
+        queryParams: { [queryParamName]: queryParamValue() },
+        queryParamsHandling: "merge",
+      })
+    );
+
+    return queryParamValue;
+  }
+}
+```
+
+```ts
+@Component({
+  selector: "app-root",
+  standalone: true,
+  template: `
+    <h1>Signals Demo</h1>
+    <p>Your first name is: {{ firstName() }}</p>
+  `,
+})
+export class AppComponent {
+  private queryParamService = inject(QueryParamService);
+
+  // Notice we only use the .get() method once
+  firstName = this.queryParamService.get("firstName");
+}
+```
+
+![Angular cancelling navigations](https://cdn.hashnode.com/res/hashnode/image/upload/v1710594428122/YXorJBf0j.gif?auto=format)
+
+Notice that when we visit the url in the browser, Angular triggers an initial navigation on start up, this initial navigation contains the query params actual values, those that are initially provided in the browser url. But, as soon as we execute the second effect inside the `queryParamService.get` method, a new empty navigation is triggered, which cancels the initial one, losing with it all query params data.
+
+This looks like a very tricky problem to solve. It would be convenient if Angular provided a way to synchronously navigate. However, this is impossible in general. Remember that we could configure things like guards or resolvers that will force the navigation to be an asynchronous operation.
+
+For this reason, we somehow need to **wait** for the current navigation to complete in order to execute the one that we have in the second effect. Fortunately, we can take advantage of `async/await` when dealing with signals, we just have to be careful so we don't get unexpected situations, pretty similar to what we discussed with `allowSignalWrites: true`.
+
+This is the final version of the service, including logic to wait for the current navigation to complete:
+
+```ts
+@Injectable({ providedIn: "root" })
+export class QueryParamService {
+  private activatedRoute = inject(ActivatedRoute);
+  private router = inject(Router);
+
+  private allQueryParams = toSignal(this.activatedRoute.queryParams);
+
+  get(queryParamName: string) {
+    const activatedRouteValue = computed(() => this.allQueryParams()?.[queryParamName]);
+    const queryParamValue = signal(activatedRouteValue());
+
+    effect(() => queryParamValue.set(activatedRouteValue()), { allowSignalWrites: true });
+
+    effect(async () => {
+      const newValueToPushInRoute = queryParamValue();
+
+      if (!this.router.navigated) return;
+      if (this.isRouterNavigating()) await this.waitUntilRouterIsIdle();
```

Comment:
**question:** what happens if `queryParamValue` has multiple new value while this Promise hasn't resolve yet. Will that start the async effect multiple times and consequently execute `this.router.navigate` multiple times? Could that lead to race conditions?